### PR TITLE
posix: abstract pthread_mutex_t as uint32_t

### PR DESCRIPTION
**Pull request by cfriedt**
3 commits
* posix: pthread: take care with pthread mutex resources
* posix: mutex: abstract pthread_mutex_t as uint32_t
* tests: posix: mutex: test to ensure there is no resource leakage

See individual commit messages for details.

### CODE REVIEW

**Comment by cfriedt** [comment]
Compliance check failure is a false positive (pthread functions return non-negative errno codes)
https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_create.html

**Comment by cfriedt** [comment]
* improved `to_posix_mutex()` (store the MSB in `pthread_mutex_t` to indicate it has been initialized)

**Comment by cfriedt** [comment]
Moving this back to draft because I have some unresolved feelings about it

**Comment by cfriedt** [comment]
> Is there a bug associated with this? I don't understand why changing posix_thread and pthread_mutex_t to uint32_t is a good thing.

This is an enhancement PR. However, there are several issues associated with POSIX and this has been outlined at length in the POSIX subsystem roadmap.

For more information, see #51211, #51771, additional issues / RFCs, and conversation in #posix channel.

Newlib and PicoLibC both use uint32_t to represent pthread_t, pthread_mutex_t, and pthread_cond_t. Similarly, glibc uses an integer type for abstraction.

PicoLibC will soon be the default for Zephyr.

The point of this change (and the preceding change of pthread_t) to uint32_t is to
* first, find a common point where the POSIX subsystem is able to meet libc's in the middle (i.e. use common abstractions)
* next, adopt Newlib's already quite mature POSIX-related headers for the internal POSIX subsystem (PicoLibC uses these as well)
* lastly use the [method outlined in the spec](https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_02) for declaring POSIX features

Ultimately, we need to be able to have the POSIX subsystem be independent of the libc. Some 3rd-party C libraries might not provide POSIX support, but we still need to be able to use the POSIX subsystem as a library along side them.

**Review by jeremybettis — DISMISSED** [review]

File: lib/posix/pthread_mutex.c [hunk 1]
```diff
@@ -22,10 +25,83 @@ static const pthread_mutexattr_t def_attr = {
...
+
+	/* Try and automatically associate a posix_mutex */
+	if (sys_bitarray_alloc(&posix_mutex_bitarray, 1, &bit) < 0) {
+		/* No mutexes left to allocate */
+		return NULL;
+	}
+
+	/* Record the associated posix_mutex in mu and mark as initialized */
+	*mu = mark_pthread_mutex_initialized(bit);
+
```
Comment (cfriedt) [inline lib/posix/pthread_mutex.c:88]: This also needs to do the same initialization that `pthread_mutex_init()` does

Comment (cfriedt) [reply lib/posix/pthread_mutex.c:89]: Done! (mostly, custom attributes are still applied in `pthread_mutex_init()`

**Comment by cfriedt** [comment]
* move mutex initialization to `to_posix_mutex()`
* use `to_posix_mutex()` instead of `get_posix_mutex()` in `cond_wait()` because passing in values initialized with `PTHREAD_MUTEX_INITIALIZER` is a valid use case.

**Comment by cfriedt** [comment]
* add test `test_posix_mutex_resource_exhausted()`
* add doxygen for `test_posix_mutex_resource_exhausted()`, `test_posix_mutex_resource_leak()`

**Review by jeremybettis — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by cfriedt** [comment]
@stephanosio - will need your superpowers to merge due to the compliance false positives.

**Comment by belussi** [comment]
Will this solve this problem: https://github.com/zephyrproject-rtos/zephyr/issues/35986

**Comment by cfriedt** [comment]
> Will this solve this problem? https://github.com/zephyrproject-rtos/zephyr/issues/35986

Yes (partially)

**Review by stephanosio — APPROVED** [review]

**Comment by belussi** [comment]
@cfriedt Good to hear that! Looks like POSIX implementation is a kind of blocker for using POSIX and C++ STL.

**Comment by cfriedt** [comment]
> @cfriedt Good to hear that! Looks like POSIX implementation is a kind of blocker for using POSIX and C++ STL. 

@belussi - the C++ STL works already and has very little dependency on POSIX (if any).

However, Newlib's (prototypical) c++11 thread support uses pthreads and requires automatic stack allocation.

That in itself is not so hard, but it really should be solved at the kernel (Zephyr) layer so that other threading implementations can simply re-use the same machinery. 

That will eventually will be fixed by #44379.

**Comment by belussi** [comment]
@PatrickM-ZS @cfriedt 
We are seeing exactly the same issue as mentioned in https://github.com/zephyrproject-rtos/zephyr/issues/35986. Conflicting types error. I couldn't find any solution yet but will try removing _pthreadtypes.h, so only STL's POSIX declaration is seen.

**Comment by PrzeBrudny** [comment]
Hi I have tried what @belussi suggested and commented conflicted typas in _pthreadtypes.h -> but failed on linking zephyr_pre0.elf with 2k+ errors:

`[build] Consolidate compiler generated dependencies of target zephyr_pre0
[build] [ 94%] Building C object zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj
[build] [ 94%] Linking CXX executable zephyr_pre0.elf
[build] /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/../../../../arm-zephyr-eabi/bin/ld.bfd: /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/arm-zephyr-eabi/lib/thumb/v7e-m/nofp/libstdc++_nano.a(atexit_arm.o): in function `__aeabi_atexit':
[build] atexit_arm.cc:(.text.__aeabi_atexit+0x0): multiple definition of `__aeabi_atexit'; arch/arch/arm/core/aarch32/libarch__arm__core__aarch32.a(__aeabi_atexit.c.obj):/workspaces/daughterboard-application/build/zephyr/arch/arch/arm/core/aarch32/../../../../../../../teltonika-tdf/platforms/zephyr/zephyr_workspace/zephyr-core/arch/arm/core/aarch32/__aeabi_atexit.c:26: first defined here
[build] /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/../../../../arm-zephyr-eabi/bin/ld.bfd: /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/arm-zephyr-eabi/lib/thumb/v7e-m/nofp/libm_nano.a(lib_a-acosl.o): in function `acosl':
[build] acosl.c:(.text.acosl+0x0): multiple definition of `acosl'; /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/arm-zephyr-eabi/lib/thumb/v7e-m/nofp/libstdc++_nano.a(math_stubs_long_double.o):math_stubs_long_double.cc:(.text.acosl+0x0): first defined here
[build] /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/../../../../arm-zephyr-eabi/bin/ld.bfd: /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/arm-zephyr-eabi/lib/thumb/v7e-m/nofp/libm_nano.a(lib_a-asinl.o): in function `asinl':
[build] asinl.c:(.text.asinl+0x0): multiple definition of `asinl'; /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/arm-zephyr-eabi/lib/thumb/v7e-m/nofp/libstdc++_nano.a(math_stubs_long_double.o):math_stubs_long_double.cc:(.text.asinl+0x0): first defined here
[build] /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/../../../../arm-zephyr-eabi/bin/ld.bfd: /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/arm-zephyr-eabi/lib/thumb/v7e-m/nofp/libm_nano.a(lib_a-atan2l.o): in function `atan2l':
[build] atan2l.c:(.text.atan2l+0x0): multiple definition of `atan2l'; /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/arm-zephyr-eabi/lib/thumb/v7e-m/nofp/libstdc++_nano.a(math_stubs_long_double.o):math_stubs_long_double.cc:(.text.atan2l+0x0): first defined here
[build] /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/../../../../arm-zephyr-eabi/bin/ld.bfd: /opt/zephyr-sdk-0.15.0/arm-zephyr-eabi/arm-zephyr-eabi/lib/thumb/v7e-m/nofp/libm_nano.a(lib_a-atanl.o): in function `atanl':` ...

I am using prj.conf flags:

`CONFIG_CPLUSPLUS=y
CONFIG_LIB_CPLUSPLUS=y
CONFIG_NEWLIB_LIBC=y`

**Comment by stephanosio** [comment]
Please open a discussion instead of commenting on old pull requests.

**Pull request merged**